src/v/cluster/node_status_provider.cc

```diff
@@ -0,0 +1,236 @@
+    vassert(ss::this_shard_id() == shard, "invoked on a wrong shard");
+
+    _metrics.clear();
+    _public_metrics.clear();
+
+    _members_table.local().unregister_members_updated_notification(
+      _members_table_notification_handle);
+    _timer.cancel();
+    return _gate.close().then([this] { return _node_connection_cache.stop(); });
+}
+
+std::optional<node_status>
+node_status_provider::get_node_status(model::node_id node) const {
+    if (node == _self) {
+        return node_status{
+          .node_id = node, .last_seen = rpc::clock_type::now()};
+    }
+
+    if (auto it = _peers_state.find(node); it != _peers_state.end()) {
+        return it->second;
+    }
+
+    return std::nullopt;
+}
+
+void node_status_provider::update_node_status(node_status_metadata&& meta) {
+    auto [it, inserted] = _discovered_peers.insert(meta.node_id);
+    if (inserted) {
+        vlog(
+          clusterlog.info, "Node {} has been discovered via RPC request", *it);
+    }
+
+    _peers_state[meta.node_id] = {
+      .node_id = meta.node_id, .last_seen = rpc::clock_type::now()};
+}
+
+void node_status_provider::tick() {
+    ssx::background = ssx::spawn_with_gate_then(_gate, [this] {
+                          return dispatch_node_status_requests().finally(
+                            [this] {
+                                if (!_gate.is_closed()) {
+                                    _timer.arm(_period());
+                                }
+                            });
+                      }).handle_exception([](const std::exception_ptr& e) {
+        vlog(
+          clusterlog.debug,
+          "Exception during node status provider tick: {}",
+          e);
+    });
+}
+
+ss::future<> node_status_provider::dispatch_node_status_requests() {
+    }
+
+    node_status_request request = {.sender_metadata = {.node_id = _self}};
+
+    return ssx::parallel_transform(
+      _discovered_peers.begin(),
+      _discovered_peers.end(),
+      [this, request = std::move(request)](auto peer_id) {
+          return do_node_status(peer_id, request);
+      });
+}
+
+ss::future<> node_status_provider::do_node_status(
+  model::node_id target, node_status_request r) {
+    auto gate_holder = _gate.hold();
+
+    auto reply = co_await send_node_status_request(target, r);
+    if (reply.has_error()) {
+        auto err = reply.error();
+        if (
+          err.category() == rpc::error_category()
+          && static_cast<rpc::errc>(err.value())
+               == rpc::errc::client_request_timeout) {
+            _stats.rpcs_timed_out += 1;
+        }
+
+        vlog(
+          clusterlog.debug,
+          "Error occurred while sending node status request: {}",
+          err.message());
+
+        co_return;
+    }
+
+    _stats.rpcs_sent += 1;
+    process_reply(std::move(reply));
+
+    co_return;
+}
+
+ss::future<result<node_status_reply>>
+node_status_provider::send_node_status_request(
+  model::node_id target, node_status_request r) {
```

Comment:
why does `do_node_status` enter the `_gate` but not some other methods like this one?

Reply:
The gate's purpose is to stop new request from coming in after stopping and to allow ongoing requests to finish while stopping. `do_node_status` is the entry point for sending one request. Since `do_node_status` is a coroutine, the gate was already entered when calling into `send_node_status_request`.

Reply:
+1 thanks